csharp/ql/src/semmle/code/csharp/AnnotatedType.qll

```diff
@@ -147,32 +243,47 @@ private int getBitMask(int bit) {
 }
 
 private int getElementTypeFlags(@has_type_annotation element) {
-  result = sum(int b | type_annotation(element, b) | b)
+  result = strictsum(int b | type_annotation(element, b) | b)
 }
 
-private int getTypeArgumentFlags(ConstructedGeneric generic, int argument) {
-  exists(generic.getTypeArgument(argument)) and
-  result = sum(int b | type_argument_annotation(generic, argument, b) | b)
+private Annotations::Nullability getTypeParameterNullability(
+  TypeParameterConstraints constraints, Type type
+) {
+  if specific_type_parameter_nullability(constraints, getTypeRef(type), _)
+  then specific_type_parameter_nullability(constraints, getTypeRef(type), result.getNullability())
+  else (
+    specific_type_parameter_constraints(constraints, type) and
+    result instanceof Annotations::NoNullability
+  )
 }
 
-private int getTypeParameterFlags(TypeParameterConstraints constraints, Type type) {
-  specific_type_parameter_annotation(constraints, getTypeRef(type), _) and
-  result = sum(int b | specific_type_parameter_annotation(constraints, getTypeRef(type), b) | b)
+private Annotations::Nullability getElementNullability(@has_type_annotation element) {
+  if type_nullability(element, _)
+  then type_nullability(element, result.getNullability())
+  else result instanceof Annotations::NoNullability
 }
 
 private newtype TAnnotatedType =
   TAnnotatedTypeNullability(Type type, Annotations::TypeAnnotations annotations) {
     Annotations::elementTypeAnnotations(_, type, annotations)
     or
-    exists(ConstructedGeneric c, int i |
-      type = c.getTypeArgument(i) and
-      annotations = Annotations::TAnnotationFlags(getTypeArgumentFlags(c, i))
+    exists(AnnotatedConstructedType c, int i |
+      type = c.getType().(ConstructedType).getTypeArgument(i) and
+      annotations = Annotations::getChild(c.getAnnotations(), i)
     )
     or
-    annotations = Annotations::TAnnotationFlags(getTypeParameterFlags(_, type))
+    annotations.getNoFlagsNullability() = getTypeParameterNullability(_, type)
+    or
+    // All types have at least one annotated type
+    annotations.getNoFlagsNullability() instanceof Annotations::NoNullability
+    or
+    exists(AnnotatedArrayType at |
+      type = at.getType().(ArrayType).getElementType() and
```

Comment:
Add
```
override ArrayType getType()
```
to the class `Annotated ArrayType` and get rid of the infix cast.

Reply:
This also introduced a negative recursion.